libgadget/fof.c

```diff
@@ -110,18 +110,20 @@ static void fof_assign_grnr(struct BaseGroup * base, const int NgroupsExt, MPI_C
 void fof_label_primary(ForceTree * tree, MPI_Comm Comm);
 extern void fof_save_particles(FOFGroups * fof, int num, int SaveParticles, MPI_Comm Comm);
 
+/* The alignment is important for the hardware atomic instruction to work*/
+#define __aint128 __int128 __attribute__ (( aligned(16)))
+
 typedef struct {
     TreeWalkQueryBase base;
     MyFloat Hsml;
-    MyIDType MinID;
+    __aint128 MinIDandTask;
```

Comment:
Use an anonymous union?

Reply:
Or perhaps define typedef union {
... uint128_t ....
struct {
  MinID,
  Task,
} FOFLabel;

Reply:
That was my first idea, but I'm not sure how to guarantee that the compiler won't pad the struct. Any change in alignment or padding would mean the hardware instruction couldn't be used and the locking could silently not be working.

Reply:
I think that gcc behaves correctly, but I don't think it is guaranteed by the standard (is it?).

Any future icc release that pads struct members would just silently introduce race conditions which I think we wouldn't notice

Reply:
Is it allowed to add the alignment (16) attribute the the union?

struct {} is all 8 bytes, so it will be tightedly packed.

Reply:
Really I have no idea! It seems like one of those standard corners where things are often wrongly implemented.